src/engine/source/builder/builders/opBuilderHelperUpgradeConfirmation.cpp

```diff
@@ -0,0 +1,108 @@
+
+namespace sint = base::utils::socketInterface;
+using helper::base::Parameter;
+
+namespace builder::internals::builders
+{
+
+// field: +upgrade_confirmation_send/ar_message
+base::Expression opBuilderHelperSendUpgradeConfirmation(const std::any& definition)
+{
+    // Extract parameters from any
+    auto [targetField, name, raw_parameters] = helper::base::extractDefinition(definition);
+    // Identify references and build JSON pointer paths
+    auto parameters {helper::base::processParameters(name, raw_parameters)};
+    // Assert expected number of parameters
+    helper::base::checkParametersSize(name, parameters, 1);
+    // Format name for the tracer
+    name = helper::base::formatHelperName(name, targetField, parameters);
+
+    // Socket instance
+    std::shared_ptr<sint::unixSecureStream> socketUC {std::make_shared<sint::unixSecureStream>(WM_UPGRADE_SOCK)};
+
+    std::string rValue {};
+    const helper::base::Parameter rightParameter {parameters[0]};
+    const auto rValueType {rightParameter.m_type};
+    rValue = rightParameter.m_value;
+
+    // Tracing
+    const auto successTrace {fmt::format("[{}] -> Success", name)};
+
+    const std::string failureTrace1 {
+        fmt::format("[{}] -> Failure: Message reference \"{}\" not found", name, parameters[0].m_value)};
+    const std::string failureTrace2 {fmt::format("[{}] -> Failure: The message is empty", name)};
+    const std::string failureTrace3 {
+        fmt::format("[{}] -> Failure: Upgrade confirmation message could not be sent", name)};
+    const std::string failureTrace4 {
+        fmt::format("[{}] -> Failure: Error trying to send upgrade confirmation message: ", name)};
+    const std::string failureTrace5 {
+        fmt::format("[{}] -> Failure: Message should be a JSON object: ", name)};
+    // Function that implements the helper
+    return base::Term<base::EngineOp>::create(
+        name,
+        [=, targetField = std::move(targetField), name = std::move(name)](
+            base::Event event) -> base::result::Result<base::Event>
+        {
+            std::string query {};
+            bool messageSent {false};
+
+            // Check if the value comes from a reference
+            if (Parameter::Type::REFERENCE == rValueType)
+            {
+                std::string resolvedRValue;
+                //Verify that its a non-empty object
+                if(event->isObject(rValue) && event->getObject(rValue).value().size())
+                {
+                    query = event->str(rValue).value();
+                }
+                else
+                {
+                    return base::result::makeFailure(event, failureTrace5);
+                }
+            }
+            else // Direct value not allowed
+            {
+                return base::result::makeFailure(event, failureTrace1);
+            }
+
+            if (query.empty())
+            {
+                return base::result::makeFailure(event, failureTrace2);
+            }
```

Comment:
```suggestion
                if(!event->isObject(rValue))
                {
                     return base::result::makeFailure(event, failureTrace5);
                }
                query = event->str(rValue).value();
            }

            if (query.empty() || query == "{}")
            {
                return base::result::makeFailure(event, failureTrace2);
            }
```

Reply:
`event->getObject(rValue).value().size()` can be expensive, it involves creating new json in a new array. we can avoid it especially if we know that we don't need that matrix.

Reply:
right, sounds good!